adoc/chapters/programming_interface.adoc

```diff
@@ -6554,10 +6554,16 @@ yet bound to a <<command-group>>.  Before such an accessor can be used in a
 <<command>>, it must be bound by calling [code]#handler::require()#.  If a
 placeholder accessor is passed as an argument to a <<command>> without first
 being bound to a <<command-group>> with [code]#handler::require()#, the
-implementation throws a synchronous [code]#exception# with the
-[code]#errc::kernel_argument# error code when the <<command>> is submitted.
+implementation may throw a synchronous [code]#exception# with the
+[code]#errc::kernel_argument# error code when the <<command>> is submitted. If
+the implementation does not throw an synchronous [code]#exception# upon
+<<command>> submission, an asynchronous [code]#exception# will be thrown during
+the execution of the <<command>> if it calls any of the member functions stating
+so (see <<sec:accessors.command.buffer.interface>> and
+<<sec:accessor.common.members>>.)
```

Comment:
How do you imagine this be implemented in the library-only case? I'm concerned that then any of these functions need to have some runtime check for every access, which could harm performance.

Reply:
I believe you're right, this would likely require some sort of runtime check for `is_placeholder() && IsThisAccessorBoundInThisCommand()`. Though we know  `IsThisAccessorBoundInThisCommand()` would be the same in each call to it inside the given command, I struggle to see how the compiler would be able to discern that. As such, I fear you're right that the implementation would effectively have to check with every call.

Do you have alternatives in mind? We could just drop the idea of throwing asynchronously for this, leaving putting the burden onto the user for cases where we can't (or don't want the overhead of) checking for unbound accessors. If we do that, we could add an interface to check if an accessor is bound, but to that end the user might as well just make sure to call require with all their accessors if they've gotten that far anyway.
Implementations can of course offer a mode that adds an asynchronous exception like what is proposed here if the user enables it. Defining such a mode seems outside the spec's responsibilities though.

Reply:
I don't really know how to handle that in a better way either. The easiest fix would likely be to adopt a "*may* throw an asynchronous exception" wording here. Generally, accessors are constructed such that once they are inside a kernel/host task little checking needs to happen for performance reasons. So I think it shouldn't be surprising to users (and maybe well-aligned with other accessor uses) that such checks may not be in place.

Reply:
After all, the exception is a service to inform the user that they're doing something they shouldn't. As such, I am okay with letting it be entirely optional for an implementation to gauge whether it is worth doing. @gmlueck - Do you feel strongly about that solution?

Reply:
Without exception thrown, how can one debug this kind of bug (aka using a placeholder without behind bound)?

Reply:
It would be possible to switch to another implementation throwing an exception during debug time, or using a pure CPU implementation with instrumentation of `operator[]` and `get_pointer()`, or use a pure CPU implementation compiled with UBsan or equivalent mode.

Reply:
If we think the error check will cause too much overhead, then we can remove the requirement.  The wording would then be much simpler:

> If a placeholder accessor is passed as an argument to a \<\<command\>\> without first
being bound to a \<\<command-group\>\> with [code]#handler::require()#, the **behavior is undefined**.

We would also remove all the proposed wording below about throwing an async exception in this case.

I think it would be good to re-examine error checking requirements in SYCL-Next.  We need some overall philosophy about error checking.  Currently, the spec is very inconsistent here.  In some cases, we require an error check and in others we just say the behavior is undefined.  I don't think we have a rationale for choosing one vs. the other.

> Without exception throwed, how can one debug this kind of bug (aka using a placeholder without behind bound)?

Even if the spec does not require an exception in this case, implementations are still allowed to do this.  I think DPC++ will continue to throw an exception in the case of kernels that use an unbound placeholder. Library only implementations could provide this ability under an compile-time option.

Reply:
Agree with greg that a coherent philosophy is required.

Also Maybe for SYCL-next we can add a new queue optional parameter, `require-throwing` or something like that?  It's always better to give control to users about the overhead /behavior they want (see discussion about returning an event) ? This can also be interesting for the SC user.